python/pyspark/sql/connect/session.py

```diff
@@ -1034,6 +1034,20 @@ def profile(self) -> Profile:
 
     profile.__doc__ = PySparkSession.profile.__doc__
 
+    def __custom_reduce_handler__(self) -> Tuple:
```

Comment:
qq: why can't we just fix `__reduce__`?

Reply:
As the comment below says, it's ony sensibel to use when the active session is correct.

Reply:
Can we simply call `SparkSession.getActiveSession`, and throw an exception if it returns `None`? We can also add an explicit session ID validation by `self._session_id == new._session_id` comparison, and would not have to bother with the custom dispatcher stuff.

Reply:
I am asking this because this ser/de logic should be able to be reused in many other places such as Python Data Source and Stored Procedure (cc @zhengruifeng and @allisonwang-db FYI)

Reply:
THat mifght be valid. The session ID might not be enough since it depends as well on how the channel is established.

Reply:
Could we start with `__reduce__`? Even when how channel created is different, I feel like the same Spark Session with the same session ID / user name should work identically with the same API provided.

If not, I would prefer to isolate those specialized logic into one place. For example, we can create something called `ConnectCloudPickleSerializer`, place `__custom_reduce_handler__`  around that file, handles `dispatch_handlers` within `ConnectCloudPickleSerializer`.

Placing `dispatch_handlers` to other serializers, fixing `SparkSession`, and `cloudpickle` seems too much to handle the specialized case.

Reply:
Ok, fine with me. Let's start with the default reduce.